src/crab/ebpf_domain.cpp

```diff
@@ -413,6 +552,10 @@ NumAbsDomain ebpf_domain_t::TypeDomain::join_by_if_else(const NumAbsDomain& inv,
     return true_case | false_case;
 }
 
+// TODO: Currently same_type() only returns true for singleton types. It should
+// be updatd to correctly return true for two registers that are known to both hold
+// either a packet or shared, such as immediately following a join of two branches
+// where both had been assigned packet in one branch and shared in the other.
```

Comment:
This comment is incorrect, and would indicate a serious bug otherwise.

Here are two test cases to demonstrate this point (and trigger the exception noted below). The idea is the existence of `r1.type=r2.type` in the first test, and its absence (producing a reasonable error message) in the second test. 
(I did not check manually the correctness of all the invariants in the `post` parts, since I believe this is beside the point)
```
test-case: same_type

pre: ["r0.type=number",
      "r8.type=ctx", "r8.ctx_offset=4",
      "r9.type=ctx", "r9.ctx_offset=0"]

code:
  <start>: |
    if r0 == 0 goto <stack>
  <number>: |
    r1 = 1
    r2 = r0
    goto <out>
  <stack>: |
    r1 = r8
    r2 = r9
    goto <out>
  <out>: |
    r3 = r2
    r3 -= r1; trigger same_type(r2, r1)

post:
  - r1.type=r2.type
  - r0.type=number
  - r1.ctx_offset=4
  - r1.type in {number, ctx}
  - r1.type-r3.type<=1
  - r2.ctx_offset=0
  - r2.type in {number, ctx}
  - r2.type-r3.type<=1
  - r3.ctx_offset=0
  - r3.type-r1.type<=0
  - r3.type-r2.type<=0
  - r3.type=number
  - r8.ctx_offset=4
  - r8.type=ctx
  - r9.ctx_offset=0
  - r9.type=ctx

---
test-case: not same_type

pre: ["r0.type=number",
      "r8.type=ctx", "r8.ctx_offset=4",
      "r9.type=ctx", "r9.ctx_offset=0"]

code:
  <start>: |
    if r0 == 0 goto <stack>
  <number>: |
    r1 = 1
    r2 = r9
    goto <out>
  <stack>: |
    r1 = r8
    r2 = r0
    goto <out>
  <out>: |
    r3 = r2
    r3 -= r1; trigger same_type(r2, r1)


post:
  - r0.type=number
  - r1.type in {number, ctx}
  - r1.ctx_offset=4
  - r2.type in {number, ctx}
  - r2.ctx_offset=0
  - r8.type=ctx
  - r8.ctx_offset=4
  - r9.type=ctx
  - r9.ctx_offset=0

messages:
  - "8: r1.type == r3.type"
```

I've added these tests to `jump.yaml` and ran them using `./run_yaml test-data/jump.yaml`. After fixing the exception, both pass. It does not mean the discussion below is resolved, since they do not test the precision I was talking about, but are intended to trigger `same_type()` and the code that use it.

Reply:
Thanks!   I've added these test cases and updated the relevant subtraction operation code to do what I think is the right thing.   I am now working on updating the addition operation code similarly.

Reply:
Updated to add and handle test cases for addition and comparison as well.   Let me know if this is not sufficient to resolve the comment, I didn't see any other remaining places.